src/lua/core/Sandbox.cpp

```diff
@@ -104,6 +105,25 @@ static int l_log_warning(lua_State *L)
 	return 0;
 }
 
+static lua_CFunction l_original_io_open = nullptr;
+
+static int l_patched_io_open(lua_State* L)
```

Comment:
Rather than validate the full path to the user's data directory, I'd prefer the path API here either deal with paths in the form `user://path/to/file` (and the corresponding `data://path/to/file` with read-only access) or relative-only paths. `FileSystem::JoinPathBelow(userdir, path)` will ensure the path is a relative path and doesn't attempt to escape the user directory.

Alternatively, wrapping `FileSource` / `FileSourceFS` and providing a different API that doesn't involve the conventions of the `io` library might be a better idea, as that would allow transparent use of the virtual filesystem utilities.

Reply:
Nice